### PR TITLE
Subclassing context.md

### PR DESCRIPTION
**Pull request by Diabolical5777**

### CODE REVIEW

File: pydis_site/apps/content/resources/guides/python-guides/subclassing_context.md
Comment (jchristgit) [inline pydis_site/apps/content/resources/guides/python-guides/subclassing_context.md]: I think it's probably better for us to link the official Python docs here (and also trim the whitespace):

```suggestion
First, the [documentation on inheritance](https://docs.python.org/3/tutorial/classes.html#inheritance) on subclassing will provide some fundamental knowledge, which is highly suggested before moving on to this topic, as subclassing [Context](https://discordpy.readthedocs.io/en/latest/ext/commands/api.html#discord.ext.commands.Context) can ultimately be a complicated task.
```

File: pydis_site/apps/content/resources/guides/python-guides/subclassing_context.md
Comment (jchristgit) [inline pydis_site/apps/content/resources/guides/python-guides/subclassing_context.md]:
```suggestion
Subclassing Context can be very beneficial as it allows you to set custom Context methods that can be used in your code. Some applications and examples are provided below.
```

File: pydis_site/apps/content/resources/guides/python-guides/subclassing_context.md
Comment (jchristgit) [inline pydis_site/apps/content/resources/guides/python-guides/subclassing_context.md]: Where will this link go?

File: pydis_site/apps/content/resources/guides/python-guides/subclassing_context.md
Comment (jchristgit) [inline pydis_site/apps/content/resources/guides/python-guides/subclassing_context.md]: Typo, let's split to two lines while we're at it:

```suggestion
# This View is for the prompt function below in the Context subclass.
# This is completely optional and its purpose is demonstrating applications of subclassing Context.
```

File: pydis_site/apps/content/resources/guides/python-guides/subclassing_context.md
Comment (jchristgit) [inline pydis_site/apps/content/resources/guides/python-guides/subclassing_context.md]: As for code style, this could probably be a docstring.

File: pydis_site/apps/content/resources/guides/python-guides/subclassing_context.md
Comment (jchristgit) [inline pydis_site/apps/content/resources/guides/python-guides/subclassing_context.md]: As for code style, this could probably be a docstring.

File: pydis_site/apps/content/resources/guides/python-guides/subclassing_context.md
Comment (jchristgit) [inline pydis_site/apps/content/resources/guides/python-guides/subclassing_context.md]: The comments here are probably better suited to be docstrings, since it's useful for the call site to know about.

File: pydis_site/apps/content/resources/guides/python-guides/subclassing_context.md
Comment (jchristgit) [inline pydis_site/apps/content/resources/guides/python-guides/subclassing_context.md]: Typo:

```suggestion
    async def send_success_embed(self, message, **kwargs):
```

File: pydis_site/apps/content/resources/guides/python-guides/subclassing_context.md
Comment (jchristgit) [inline pydis_site/apps/content/resources/guides/python-guides/subclassing_context.md]: This docstring is kind of mangled up. Did you mean this?

```suggestion
    async def prompt(self,message: str, *, timeout: float = 60.0, delete_after: bool = True, author_id: Optional[int] = None,) -> Optional[bool]:
        """
        Parameters
        ----------
        message: str
            The message to show along with the prompt.
        timeout: float (Defaults to 60 seconds)
            How long to wait before stopping the view
        delete_after: bool (Defaults to True)
            Whether to delete the confirmation message after we're done.
        author_id: Optional[int] (Defaults to ctx.author)
            The member who should respond to the prompt.

        Returns  
        -------
        True if the confirm button is pressed,
        False if cancel button is pressed,
        None if automatically denied due to no response
        """
```

File: pydis_site/apps/content/resources/guides/python-guides/subclassing_context.md
Comment (jchristgit) [inline pydis_site/apps/content/resources/guides/python-guides/subclassing_context.md]: We're not passing anything here, and I don't think we need to override `__init__` in the first place, do we?

File: pydis_site/apps/content/resources/guides/python-guides/subclassing_context.md
Comment (jchristgit) [inline pydis_site/apps/content/resources/guides/python-guides/subclassing_context.md]: Let's keep code style consistent:

```suggestion
    await ctx.send_success_embed('This is another example of a subclassed embed!', title='Subclassing Embed', timestamp=datetime.datetime.utcnow())
```

File: pydis_site/apps/content/resources/guides/python-guides/subclassing_context.md
Comment (jchristgit) [inline pydis_site/apps/content/resources/guides/python-guides/subclassing_context.md]: This is invalid due to `elife`, and we should use more Pythonic checks:

```suggestion
    if request_confirm:
        # The user confirmed so we can continue with the command

    else:
        # The user cancelled so we can return 
```

File: pydis_site/apps/content/resources/guides/python-guides/subclassing_context.md
Comment (jchristgit) [inline pydis_site/apps/content/resources/guides/python-guides/subclassing_context.md]: It may be nice to maybe run the full codeblock through a formatter to adapt it to a single style.

**Review by jchristgit — CHANGES_REQUESTED** [review]

File: pydis_site/apps/content/resources/guides/python-guides/subclassing_context.md
Comment (Tenuka-R) [reply pydis_site/apps/content/resources/guides/python-guides/subclassing_context.md]: I put a comment of `key word arguments` there in the init with the intention of increasing the need to go over the subclassing_bot article to understand how to subclass bot. If I were to add the key word arguments, in my opinion people are more likely to just copy paste without fully understanding (in line 12, I wanted to link the subclassing_bot file). But if this isn't a good reasoning I can change it.

**Review by Shom770 — APPROVED** [review]
Looks good to me! Though before this gets merged, you could implement some of Volccy's feedback.

**Comment by Tenuka-R** [comment]
> Looks good to me! Though before this gets merged, you could implement some of Volccy's feedback.

Which ones? I thought I implemented them all

File: pydis_site/apps/content/resources/guides/python-guides/subclassing_context.md
Comment (Bluenix2) [inline pydis_site/apps/content/resources/guides/python-guides/subclassing_context.md]: Isn't it easier to only mention subclassing the Bot class? You don't need to link to that first documentation, since that is done in the bot one.

Comment (Bluenix2) [inline pydis_site/apps/content/resources/guides/python-guides/subclassing_context.md]: This example, was in-fact, not very short 😅

File: pydis_site/apps/content/resources/guides/python-guides/subclassing_context.md
Comment (Bluenix2) [inline pydis_site/apps/content/resources/guides/python-guides/subclassing_context.md]: This line is waaaaaaaaaaaaaaayyyyyyyy too long

File: pydis_site/apps/content/resources/guides/python-guides/subclassing_context.md
Comment (Bluenix2) [reply pydis_site/apps/content/resources/guides/python-guides/subclassing_context.md]: I agree, go through this with a linter. It has some inconsistencies and odd-ness.

Please capitalize comments and place them either on the same line, or above, the code.

File: pydis_site/apps/content/resources/guides/python-guides/subclassing_context.md
Comment (Bluenix2) [inline pydis_site/apps/content/resources/guides/python-guides/subclassing_context.md]: This is invalid syntax, if you wish to have the user fill in something please add `...`

File: pydis_site/apps/content/resources/guides/python-guides/subclassing_context.md
Comment (Bluenix2) [inline pydis_site/apps/content/resources/guides/python-guides/subclassing_context.md]: The focus is meant to be on the context, yet this is placed first and is probably longer than the actual context subclass.

Can we consider another type of example?

File: pydis_site/apps/content/resources/guides/python-guides/subclassing_context.md
Comment (Bluenix2) [inline pydis_site/apps/content/resources/guides/python-guides/subclassing_context.md]: This would do great as a docstring.

File: pydis_site/apps/content/resources/guides/python-guides/subclassing_context.md
Comment (Bluenix2) [inline pydis_site/apps/content/resources/guides/python-guides/subclassing_context.md]: This docstring is missing a description other than the reference of parameters.

Additionally, the typings are present in the definition so you don't need to repeat them in the docstring.

File: pydis_site/apps/content/resources/guides/python-guides/subclassing_context.md
Comment (Bluenix2) [inline pydis_site/apps/content/resources/guides/python-guides/subclassing_context.md]: I am gonna be honest, this guide probably isn't very helpful to most. It only has a missleading example.

I think what we can do, to improve this, is add a full text-explanation of the steps the user has to go through. This can either take the shape of maybe an ordered list with smaller codeblocks of the entire example, or perhaps the full codeblock but with a bunch of comments... or something else I can't think of.

Structuring it like that would be much more helpful to the user!

**Comment by Shom770** [comment]
> > Looks good to me! Though before this gets merged, you could implement some of Volccy's feedback.
> 
> Which ones? I thought I implemented them all

I assumed since some of them weren't marked as resolved yet that they weren't completed, a mistake on my end haha

**Comment by Bluenix2** [comment]
As per [this message in the discord](https://discord.com/channels/267624335836053506/635950537262759947/966658073009725542) I will be taking over this PR.

**Comment by Xithrius** [comment]
Hey @Bluenix2, what's your status on this PR?

Thanks!

**Comment by Bluenix2** [comment]
I do not have any other changes - this is ready for review! The preview failed because of ratelimits, I am not sure whether I can re-run them.

**Comment by Xithrius** [comment]
Very cool. Thanks!

File: pydis_site/apps/content/resources/guides/python-guides/subclassing_context.md
Comment (Robin5605) [inline pydis_site/apps/content/resources/guides/python-guides/subclassing_context.md]: Could we perhaps move this into it's own method? I feel like nested functions and using nonlocal just makes it more confusing and unpythonic.

File: pydis_site/apps/content/resources/guides/python-guides/subclassing_context.md
Comment (Robin5605) [inline pydis_site/apps/content/resources/guides/python-guides/subclassing_context.md]: I think it would be better if we were to extract this part out so it's clearer.
Also the prompt should have "members" at the end, other wise it just looks like:
`Are you sure you want to ban 5?` which doesn't make much sense.

```suggestion
    prompt = await ctx.prompt(f"Are you sure you want to ban {len(members)} members?")
    if not prompt:
```

File: pydis_site/apps/content/resources/guides/python-guides/subclassing_context.md
Comment (Robin5605) [inline pydis_site/apps/content/resources/guides/python-guides/subclassing_context.md]: I believe this should work and looks better.
```suggestion
        for reaction in ('✅', '❌'):
```

File: pydis_site/apps/content/resources/guides/python-guides/subclassing_context.md
Comment (Robin5605) [inline pydis_site/apps/content/resources/guides/python-guides/subclassing_context.md]: I want to include enabling message content as I see this to be one of the biggest issues currently.
```suggestion
# This makes sure your bot receives message content and run command handlers appropriately.
# Without this, our commands wouldn't work and we wouldn't get any errors either. 
intents = discord.Intents.default()
intents.message_content = True

bot = CustomBot(...)  # Replace with the arguments for the bot
```

**Review by Robin5605 — CHANGES_REQUESTED** [review]
Minors changes with the exception of extracting the `check` function into it's own method. Thanks for the PR!

File: pydis_site/apps/content/resources/guides/python-guides/subclassing_context.md
Comment (shtlrs) [inline pydis_site/apps/content/resources/guides/python-guides/subclassing_context.md]: Nit: customizability *

File: pydis_site/apps/content/resources/guides/python-guides/subclassing_context.md
Comment (shtlrs) [reply pydis_site/apps/content/resources/guides/python-guides/subclassing_context.md]: I agree with Ronin here, if you want your method to be "protected" or "private" then I'd suggest to prefix it with either a "_" or a "__"
Example:

`def _protected_method(...)` **OR** `def __private_method(...)`

File: pydis_site/apps/content/resources/guides/python-guides/subclassing_context.md
Comment (shtlrs) [inline pydis_site/apps/content/resources/guides/python-guides/subclassing_context.md]:
```suggestion
async def mass_ban(ctx: CustomContext, members: commands.Greedy[discord.Member]):
```

Comment (shtlrs) [reply pydis_site/apps/content/resources/guides/python-guides/subclassing_context.md]: And I don't see the use of `Greedy` if you have no other parameters after it, so I think a list of discord.Member will work as much

**Review by shtlrs — DISMISSED** [review]
Looks good ! Just a couple of tiny comments.

**Comment by Bluenix2** [comment]
Thanks for these reviews! I am hoping to get to this soon, sorry

**Comment by Robin5605** [comment]
Hey @Bluenix2 ! Don't mean to rush you, but what's your status on getting the reviews resolved? If you're unable to, we can make this up for grabs for someone else, no worries!
Thanks 😄

File: pydis_site/apps/content/resources/guides/python-guides/subclassing_context.md
Comment (Bluenix2) [reply pydis_site/apps/content/resources/guides/python-guides/subclassing_context.md]: This function is a closure which closes over `msg`, `confirmation` and `self`. We need access to these during the check.

One possibility *would* be putting this as a method and then using `functools.partial` however I think that a closure is more clear.

File: pydis_site/apps/content/resources/guides/python-guides/subclassing_context.md
Comment (Bluenix2) [reply pydis_site/apps/content/resources/guides/python-guides/subclassing_context.md]: This was a conscious decision, as it is generally seen as bad practice to store unicode characters in source files: they are quite difficult to work with. The alternative used here is the most readable one.

That said, I will change this since it appears that I have been outvoted.

**Review by wookie184 — APPROVED** [review]
Seems good enough to merge.

**Comment by wookie184** [comment]
wanted to rename the file for consistency, will need someone elses approval

File: pydis_site/apps/content/resources/guides/python-guides/discordpy-subclassing-context.md
Comment (shtlrs) [inline pydis_site/apps/content/resources/guides/python-guides/discordpy-subclassing-context.md]:
```suggestion
adding whatever functionality you wish. Usually this is done by adding custom methods or properties, so that you don't need to
```

File: pydis_site/apps/content/resources/guides/python-guides/discordpy-subclassing-context.md
Comment (shtlrs) [inline pydis_site/apps/content/resources/guides/python-guides/discordpy-subclassing-context.md:43]: Shouldn't this be: "confirmed, denied or didn't interact" ?

File: pydis_site/apps/content/resources/guides/python-guides/discordpy-subclassing-context.md
Comment (shtlrs) [inline pydis_site/apps/content/resources/guides/python-guides/discordpy-subclassing-context.md]: I think it's better to stick with one representation of these reactions.
For example, we can check if emoji == '✅' or '❌'

File: pydis_site/apps/content/resources/guides/python-guides/discordpy-subclassing-context.md
Comment (shtlrs) [inline pydis_site/apps/content/resources/guides/python-guides/discordpy-subclassing-context.md:112]: Maybe refer to the intents tag here for more information/details ?

**Review by shtlrs — CHANGES_REQUESTED** [review]
Looks good overall, just some final touches.

File: pydis_site/apps/content/resources/guides/python-guides/discordpy-subclassing-context.md
Comment (Bluenix2) [reply pydis_site/apps/content/resources/guides/python-guides/discordpy-subclassing-context.md:112]: Good idea, however, I am not sure the best way to do this considering it is not on our site.

Comment (wookie184) [reply pydis_site/apps/content/resources/guides/python-guides/discordpy-subclassing-context.md:112]: Since it's not that relevant to the actual aim of the guide I think it's fine to leave it as is personally.

**Review by jchristgit — APPROVED** [review]
:shipit::shipit::shipit::shipit:

Thank you!

**Review by Xithrius — APPROVED** [review]

**Pull request merged**
